src/Data/Bin/Ordering.agda

```diff
@@ -0,0 +1,790 @@
+
+≰⇒> :  _≰_ ⇒ _>_
+≰⇒> {x} {y} x≰y  with <-cmp x y
+... | tri< lt _  _   =  contradiction (<⇒≤ {x} {y} lt) x≰y
+... | tri≈ _  eq _   =  contradiction (≤-reflexive eq) x≰y
+... | tri> _  _  x>y =  x>y
+
+≤,≢⇒< :  ∀ {x y} → x ≤ y → x ≢ y → x < y
+≤,≢⇒< (inj₁ x<y) _   =  x<y
+≤,≢⇒< (inj₂ x≡y) x≢y =  contradiction x≡y x≢y
+
+<-≤-trans :  ∀ {x y z} → x < y → y ≤ z → x < z
+<-≤-trans             x<y (inj₁ y<z) =  <-trans x<y y<z
+<-≤-trans {x} {_} {_} x<y (inj₂ y≡z) =  subst (x <_) y≡z x<y
+
+≤-<-trans :  ∀ {x y z} → x ≤ y → y < z → x < z
+≤-<-trans             (inj₁ x<y) y<z =  <-trans x<y y<z
+≤-<-trans {_} {_} {z} (inj₂ x≡y) y<z =  subst (_< z) (sym x≡y) y<z
+
+≡0-by≤ :  ∀ {x y} → x ≤ y → y ≡ zero → x ≡ zero
+≡0-by≤ {x} {y} x≤y y≡0 =  ≤0⇒≡ x≤0
+  where
+  x≤0 = subst (x ≤_) y≡0 x≤y
+
+≢0-by≤ :  ∀ {x y} → x ≤ y → x ≢ zero → y ≢ zero
+≢0-by≤ x≤y x≢0 =  x≢0 ∘ ≡0-by≤ x≤y
+
+_<?_ : Decidable _<_
+x <? y  with <-cmp x y
+... | tri< lt  _ _ =  yes lt
+... | tri≈ ¬lt _ _ =  no ¬lt
+... | tri> ¬lt _ _ =  no ¬lt
+
+_≤?_ : Decidable _≤_
+x ≤? y  with <-cmp x y
+... | tri< lt _  _  =  yes (<⇒≤ {x} {y} lt)
+... | tri≈ _  eq _  =  yes (≤-reflexive eq)
+... | tri> _  _  gt =  no (<⇒≱ {y} {x} gt)
+
+≤-isPreorder :  IsPreorder _≡_ _≤_
+≤-isPreorder =  record
+  { isEquivalence = isEquivalence
+  ; reflexive     = ≤-reflexive
+  ; trans         = ≤-trans
+  }
+
+<-respˡ-≈ :  _<_ Respectsˡ _≡_
+<-respˡ-≈ {x} {_} =  subst (_< x)
+
+<-respʳ-≈ :  _<_ Respectsʳ _≡_
+<-respʳ-≈ {x} {_} =  subst (x <_)
+
+<-resp-≈ :  _<_ Respects₂ _≡_
+<-resp-≈ =  ( <-respʳ-≈ , <-respˡ-≈ )
+
+------------------------------------------------------------------------------
+-- Several monotonicity proofs for _+_, suc
+------------------------------------------------------------------------------
+
+module IneqReasoning = InequalityReasoning {A = Bin} {_≈_ = _≡_} ≤-isPreorder <-trans
+                                           <-resp-≈ <⇒≤ <-≤-trans ≤-<-trans
+
++-mono-≤ :  _+_ Preserves₂ _≤_ ⟶ _≤_ ⟶ _≤_
++-mono-≤ {x} {x'} {y} {y'} x≤x' y≤y' =  begin
+  x + y                  ≈⟨ sym $ cong₂ _+_ (fromℕ-toℕ x) (fromℕ-toℕ y) ⟩
+  fromℕ m + fromℕ n      ≈⟨ sym (fromℕ-homo-+ m n) ⟩
+  fromℕ (m ℕ.+ n)        ≤⟨ fromℕ-mono-≤ (ℕₚ.+-mono-≤ m≤m' n≤n') ⟩
+  fromℕ (m' ℕ.+ n')      ≈⟨ fromℕ-homo-+ m' n' ⟩
+  fromℕ m' + fromℕ n'    ≈⟨ cong₂ _+_ (fromℕ-toℕ x') (fromℕ-toℕ y') ⟩
+  x' + y'                ∎
+  where
+  open IneqReasoning
+  m    = toℕ x;             m'   = toℕ x'
+  n    = toℕ y;             n'   = toℕ y'
+  m≤m' = toℕ-mono-≤ x≤x';   n≤n' = toℕ-mono-≤ y≤y'
+
++-monoˡ-≤ :  ∀ x → (_+ x) Preserves _≤_ ⟶ _≤_
++-monoˡ-≤ x {y} {z} y≤z =  +-mono-≤ {y} {z} {x} {x} y≤z (≤-refl {x})
+
++-monoʳ-≤ :  ∀ x → (x +_) Preserves _≤_ ⟶ _≤_
++-monoʳ-≤ x {y} {z} y≤z =  +-mono-≤ {x} {x} {y} {z} (≤-refl {x}) y≤z
+
++-mono-<-≤ :  _+_ Preserves₂ _<_ ⟶ _≤_ ⟶ _<_
++-mono-<-≤ {x} {x'} {y} {y'} x<x' y≤y' =  begin-strict
+    x + y                     ≈⟨ sym $ cong₂ _+_ (fromℕ-toℕ x) (fromℕ-toℕ y) ⟩
```

Comment:
Please could you better align this reasoning proof

Reply:
Done